src/node_api.cc

```diff
@@ -0,0 +1,2430 @@
+}
+
+napi_status napi_unwrap(napi_env e, napi_value jsObject, void** result) {
+  // Omit NAPI_PREAMBLE and GET_RETURN_STATUS because V8 calls here cannot throw
+  // JS exceptions.
+  CHECK_ARG(jsObject);
+  CHECK_ARG(result);
+
+  v8::Local<v8::Object> obj =
+      v8impl::V8LocalValueFromJsValue(jsObject).As<v8::Object>();
+
+  // Only objects that were created from a NAPI constructor's prototype
+  // via napi_define_class() can be (un)wrapped.
+  RETURN_STATUS_IF_FALSE(obj->InternalFieldCount() > 0, napi_invalid_arg);
+
+  *result = obj->GetAlignedPointerFromInternalField(0);
+
+  return napi_ok;
+}
+
+napi_status napi_create_external(napi_env e,
+                                 void* data,
+                                 napi_finalize finalize_cb,
+                                 void* finalize_hint,
+                                 napi_value* result) {
+  NAPI_PREAMBLE(e);
+  CHECK_ARG(result);
+
+  v8::Isolate* isolate = v8impl::V8IsolateFromJsEnv(e);
+
+  v8::Local<v8::Value> externalValue = v8::External::New(isolate, data);
+
+  // The Reference object will delete itself after invoking the finalizer
+  // callback.
+  new v8impl::Reference(isolate,
+      externalValue,
+      0,
+      true,
+      finalize_cb,
+      data,
+      finalize_hint);
+
+  *result = v8impl::JsValueFromV8LocalValue(externalValue);
+
+  return GET_RETURN_STATUS();
+}
+
+napi_status napi_get_value_external(napi_env e,
+                                    napi_value value,
+                                    void** result) {
+  NAPI_PREAMBLE(e);
+  CHECK_ARG(value);
+  CHECK_ARG(result);
+
+  v8::Local<v8::Value> val = v8impl::V8LocalValueFromJsValue(value);
+  RETURN_STATUS_IF_FALSE(val->IsExternal(), napi_invalid_arg);
+
+  v8::Local<v8::External> externalValue = val.As<v8::External>();
+  *result = externalValue->Value();
+
+  return GET_RETURN_STATUS();
+}
+
+// Set initial_refcount to 0 for a weak reference, >0 for a strong reference.
+napi_status napi_create_reference(napi_env e,
+                                  napi_value value,
+                                  int initial_refcount,
+                                  napi_ref* result) {
+  NAPI_PREAMBLE(e);
+  CHECK_ARG(result);
+  RETURN_STATUS_IF_FALSE(initial_refcount >= 0, napi_invalid_arg);
+
+  v8::Isolate* isolate = v8impl::V8IsolateFromJsEnv(e);
+
+  v8impl::Reference* reference = new v8impl::Reference(
+      isolate, v8impl::V8LocalValueFromJsValue(value), initial_refcount, false);
+
+  *result = reinterpret_cast<napi_ref>(reference);
+  return GET_RETURN_STATUS();
+}
+
+// Deletes a reference. The referenced value is released, and may be GC'd unless
+// there
+// are other references to it.
+napi_status napi_delete_reference(napi_env e, napi_ref ref) {
+  // Omit NAPI_PREAMBLE and GET_RETURN_STATUS because V8 calls here cannot throw
+  // JS exceptions.
+  CHECK_ARG(ref);
+
+  v8impl::Reference* reference = reinterpret_cast<v8impl::Reference*>(ref);
+  delete reference;
```

Comment:
(re: my comment above … maybe also make the `Reference` destructor private and add a `Delete` static? I know it might seem a bit superfluous but I think it would really help with readability)